Brokerages/Bitfinex/BitfinexBrokerage.cs

```diff
@@ -284,31 +284,51 @@ public override IEnumerable<BaseData> GetHistory(Data.HistoryRequest request)
+            }
+
             string resolution = ConvertResolution(request.Resolution);
-            long resolutionInMS = (long)request.Resolution.ToTimeSpan().TotalMilliseconds;
+            long resolutionInMsec = (long)request.Resolution.ToTimeSpan().TotalMilliseconds;
             string symbol = _symbolMapper.GetBrokerageSymbol(request.Symbol);
-            long startMTS = (long)Time.DateTimeToUnixTimeStamp(request.StartTimeUtc) * 1000;
-            long endMTS = (long)Time.DateTimeToUnixTimeStamp(request.EndTimeUtc) * 1000;
+            long startMsec = (long)Time.DateTimeToUnixTimeStamp(request.StartTimeUtc) * 1000;
+            long endMsec = (long)Time.DateTimeToUnixTimeStamp(request.EndTimeUtc) * 1000;
             string endpoint = $"v2/candles/trade:{resolution}:t{symbol}/hist?limit=1000&sort=1";
+            var period = request.Resolution.ToTimeSpan();
 
-            while ((endMTS - startMTS) > resolutionInMS)
+            do
             {
-                var timeframe = $"&start={startMTS}&end={endMTS}";
+                var timeframe = $"&start={startMsec}&end={endMsec}";
 
                 var restRequest = new RestRequest(endpoint + timeframe, Method.GET);
                 var response = ExecuteRestRequest(restRequest);
 
                 if (response.StatusCode != HttpStatusCode.OK)
                 {
-                    throw new Exception($"BitfinexBrokerage.GetHistory: request failed: [{(int)response.StatusCode}] {response.StatusDescription}, Content: {response.Content}, ErrorMessage: {response.ErrorMessage}");
+                    throw new Exception(
+                        $"BitfinexBrokerage.GetHistory: request failed: [{(int) response.StatusCode}] {response.StatusDescription}, " +
+                        $"Content: {response.Content}, ErrorMessage: {response.ErrorMessage}");
                 }
 
                 var candles = JsonConvert.DeserializeObject<object[][]>(response.Content)
                     .Select(entries => new Messages.Candle(entries))
                     .ToList();
 
-                startMTS = candles.Last().Timestamp + resolutionInMS;
-                var period = request.Resolution.ToTimeSpan();
+                // bitfinex exchange may return us an empty result - if we request data for a small time interval
+                // during which no trades occurred - so it's retional to ensure 'candles' list is not empty before
+                // we proceed to avoid an exception to be thrown
+                if (candles.Any())
+                {
+                    startMsec = candles.Last().Timestamp + resolutionInMsec;
+                }
+                else
+                {
+                    Log.Error($"BitfinexBrokerage.GetHistory(): Exchange returned no data for {symbol}" +
+                              $"on history request from {request.StartTimeUtc:s} to {request.EndTimeUtc:s}");
+                    break;
+                }
```

Comment:
Good catch!
Depending on the requested range an empty list can be a valid response, as noted in your comment.
If we want to log this condition, I think this message should not be logged as an error, maybe better use `Log.Trace` instead?

Reply:
Since this would be responding to a user's request for history, I think it makes sense to raise a `BrokerageMessageEvent` so the algorithm can handle it.